### PR TITLE
Improvements around nanmean, mean and nansum

### PR DESCRIPTION
**Pull request by OwenArnold**
Remaining two items marked on#1451

* On Testing
  * Wider use of TYPED_TEST for existing test cases.
  * Common TestSuite for Variable and Dataset testing
  * A fix for TYPED_TEST macro related warnings
* Nansum will work on ints directly, this allows us to remove some of our bespoke considerations with the `nanmean` implementation.

Note that I'm using GTEST_SKIP here, which will show up in the test overview (much like we have anyway for our python tests). We could avoid the explicit skip marking if that style was preferred.

### CODE REVIEW

File: core/include/scipp/core/element/arithmetic.h [hunk 1]
```diff
@@ -22,19 +22,26 @@ constexpr auto add_inplace_types =
              std::tuple<float, int64_t>, std::tuple<float, int32_t>,
              std::tuple<int64_t, bool>>;
 
-constexpr auto add_inplace_nan_types =
-    arg_list<double, float, std::tuple<double, float>,
-             std::tuple<float, double>>;
+template <class T> struct ValueType { using value_type = T; };
+template <class T> struct ValueType<ValueAndVariance<T>> {
+  using value_type = T;
+};
```
Comment (SimonHeybrock) [inline core/include/scipp/core/element/arithmetic.h:28]: What is wrong with using https://github.com/scipp/scipp/blob/b6fa0b96c739055a49d99cce34c6fd11783cacc5/core/include/scipp/core/value_and_variance.h#L22?

File: dataset/reduction.cpp [hunk 2]
```diff
@@ -83,9 +83,10 @@ DataArray nanmean(const DataArrayConstView &a, const Dim dim) {
 }
 
 DataArray nanmean(const DataArrayConstView &a) {
-  if (isInt(a.data().dtype()))
-    return mean(a);
-  return nansum(a) * reciprocal(astype(sum(isfinite(a)), a.dtype()));
+  if (isInt(a.dtype()))
+    return nansum(a) * reciprocal(astype(sum(isfinite(a)), dtype<double>));
+  else
+    return nansum(a) * reciprocal(astype(sum(isfinite(a)), a.dtype()));
```
Comment (SimonHeybrock) [inline dataset/reduction.cpp:89]: Why this change to a more complicated code?

File: test/test_nans.h [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#pragma once
+#include <gtest/gtest.h>
+#include <numeric>
+#include <type_traits>
+
+namespace {
+template <typename T> struct MeanTest : public ::testing::Test {
```
Comment (SimonHeybrock) [inline test/test_nans.h:7]: `test/` contains generic test helper, I don't think we should add something so specific here?

File: variable/reduction.cpp [hunk 4]
```diff
@@ -89,7 +89,10 @@ Variable nanmean_impl(const VariableConstView &var, const Dim dim,
                       const VariableConstView &count) {
   auto summed = nansum(var, dim);
   auto scale = reciprocal(astype(count, core::dtype<double>));
-  summed *= scale;
+  if (isInt(var.dtype()))
+    summed = summed * scale;
+  else
+    summed *= scale;
```
Comment (SimonHeybrock) [inline variable/reduction.cpp:95]:
```cpp
if (isInt(var.dtype()))
  return mean_impl(var, dim, count)
```
?

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]

File: core/include/scipp/core/element/arithmetic.h [hunk 5]
```diff
@@ -22,19 +22,26 @@ constexpr auto add_inplace_types =
...
 
 constexpr auto plus_equals =
     overloaded{add_inplace_types, [](auto &&a, const auto &b) { a += b; }};
 constexpr auto nan_plus_equals =
-    overloaded{add_inplace_nan_types, [](auto &&a, const auto &b) {
-                 using std::isnan;
-                 if (isnan(a))
-                   a = std::decay_t<decltype(a)>{0}; // Force zero
-                 if (!isnan(b))
+    overloaded{add_inplace_types, [](auto &&a, const auto &b) {
```
Comment (Matthew-Andrew) [inline core/include/scipp/core/element/arithmetic.h:33]: I think you need to check that both a and b are floating point types for this overload as isnan is called on both of them. Alternatively could remove `std::tuple<double, int64_t>, std::tuple<double, int32_t>,
             std::tuple<float, int64_t>, std::tuple<float, int32_t>,` from the allowed input types.

Comment (Matthew-Andrew) [reply core/include/scipp/core/element/arithmetic.h:33]: It may turn out to wrap std:isnan to Integral types on windows in this case.

File: dataset/reduction.cpp (hunk 2 above)
Comment (OwenArnold) [reply dataset/reduction.cpp]: Reversed change

File: core/include/scipp/core/element/arithmetic.h (hunk 1 above)
Comment (OwenArnold) [reply core/include/scipp/core/element/arithmetic.h:28]: see usage below - templated class and specialisation allow us to determine if we are dealing with underlying floating point types whether value or ValueAndVariance.

File: test/test_nans.h (hunk 3 above)
Comment (OwenArnold) [reply test/test_nans.h:7]: I had reservations, but it's used in two test suites (dataset and variable) and I wanted to avoid copying.

File: variable/reduction.cpp (hunk 4 above)
Comment (OwenArnold) [reply variable/reduction.cpp]: Changed

File: core/include/scipp/core/element/arithmetic.h (hunk 1 above)
Comment (SimonHeybrock) [reply core/include/scipp/core/element/arithmetic.h:28]: I see. Have you considered doing what you did for other functions and check for ints:
```cpp
if (isInt(var.dtype())
  return sum(var);
return transform(var, nanplus);
```
or something like that?

File: core/include/scipp/core/element/arithmetic.h (hunk 5 above)
Comment (SimonHeybrock) [reply core/include/scipp/core/element/arithmetic.h]: > It may turn out to be easier to wrap std:isnan to handle Integral types on windows in this case.

Yes, woulnd't that simplify quite a few things?

Comment (OwenArnold) [reply core/include/scipp/core/element/arithmetic.h]: > Yes, woulnd't that simplify quite a few things? There is not even need for wrapping. Just define isnan(ints) in the namespace where the kernel is defined, and it should be found with ADL?

It would also need to be used internally with the overload in `ValueAndVariance`. - But it's a good suggestion to maybe have our own special functions at the basic type level which we use directly instead of those from `std::`

File: core/include/scipp/core/element/arithmetic.h (hunk 1 above)
Comment (OwenArnold) [reply core/include/scipp/core/element/arithmetic.h:28]: With Matt's suggestion below some `detail::isnan` supporting ints and floats we could simplify all the type handling here.

Comment (SimonHeybrock) [reply core/include/scipp/core/element/arithmetic.h:29]: Now unused?

**Review by SimonHeybrock — APPROVED** [review]

**Review by SimonHeybrock — APPROVED** [review]
